packages/vite/src/node/config.ts

```diff
@@ -815,22 +819,20 @@ async function bundleConfigFile(
         }
       },
       {
-        name: 'replace-import-meta',
+        name: 'inject-file-scope-variables',
         setup(build) {
           build.onLoad({ filter: /\.[jt]s$/ }, async (args) => {
             const contents = await fs.promises.readFile(args.path, 'utf8')
+            const injectValues =
+              `const __dirname = ${JSON.stringify(path.dirname(args.path))};` +
+              `const __filename = ${JSON.stringify(args.path)};` +
```

Comment:
Taking from [tsup's page](https://github.com/egoist/tsup/blob/b9cd8d5f0973f94d494b8b0d0e65121e5712ee49/src/esbuild/index.ts#L214-L222), maybe we could use esbuild's [inject](https://esbuild.github.io/api/#inject) option to better replace this without caveats?

I haven't dug too deep, but if otherwise this PR looks good too.

Reply:
If `inject` is used, `__filename` / `__dirname` becomes the value after bundle. But here the value before bundle is expected.

For example, when `inject` is used,
```ts
// foo.ts
import { bar } from './bar';

console.log(__dirname);
console.log(bar);

// bar.ts
export const bar = __dirname;

// injectee.js
import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export { __filename, __dirname };
```
this becomes:
```js
import { fileURLToPath } from "url";
import path from "path";
var __filename = fileURLToPath(import.meta.url);
var __dirname = path.dirname(__filename);

var bar = __dirname;

console.log(__dirname);
console.log(bar); // same value with __dirname
```

Reply:
Ah you're correct. Looks like esbuild's inject accepts a file, and that can't be changed to the value before bundle. Scrap my idea then 😅